File: lab6/python/simple_test_graph.py
import sys

import general
from hashmap import hashmap_t
from graph import graph_t
import weight
from sp_algorithms import bfs, bellman_ford, dijkstra, astar_search
from shortest_path import sssp_result_t, print_path

def msg0(s):
    general.msg(0, s)
def msg1(s):
    general.msg(1, s)
    
    
def compute_path(algo, src, dst):
    s = int(src)
    d = int(dst)
    g = graph_t(10)
    g.graph_add_edge(0, weight.weight_t(1), 1)
    g.graph_add_edge(1, weight.weight_t(2), 2)
    g.graph_add_edge(2, weight.weight_t(3), 3)
    g.graph_add_edge(3, weight.weight_t(1), 4)
    g.graph_add_edge(5, weight.weight_t(1), 6)
    g.graph_add_edge(6, weight.weight_t(1), 8)
    g.graph_add_edge(8, weight.weight_t(2), 9)
    g.graph_add_edge(5, weight.weight_t(9), 9)
    if (algo == "bellman-ford-neg"):
        g.graph_add_edge(3, weight.weight_t(-1), 0)
        g.graph_add_edge(5, weight.weight_t(-5), 7)
    else:
        g.graph_add_edge(3, weight.weight_t(4), 0)
        g.graph_add_edge(5, weight.weight_t(5), 7)
        
    
    h = [None]*g.graph_get_num_nodes()
    for u in range(0, g.graph_get_num_nodes()):
        h[u] = weight.weight_t(abs(d - u))
    
    if (algo == "bellman-ford"):
        bellman_ford(g,s).sssp_to_sp_result(d).print_sp_result(sys.stdout)
    elif (algo == "bellman-ford-neg"):
        bellman_ford(g,s).sssp_to_sp_result(d).print_sp_result(sys.stdout)
    elif (algo == "dijkstra"):
        dijkstra(g,s,d).sssp_to_sp_result(d).print_sp_result(sys.stdout)
    elif (algo == "astar"):
        astar_search(g,s,d,h).print_sp_result(sys.stdout)
    elif (algo == "bfs"):
        bfs(g,s,d).sssp_to_sp_result(d).print_sp_result(sys.stdout)
    else:
        general.error("Invalid algorithm name: %s" % algo)
    
if (len(sys.argv) == 4):
    compute_path(sys.argv[1], sys.argv[2], sys.argv[3])
elif (len(sys.argv) == 5):
    if (sys.argv[4] == "-v"):
        general.set_msg_verb(0)
    elif (sys.argv[4] == "-vv"):
        general.set_msg_verb(1)
    elif (sys.argv[4] == "-vvv"):
        general.set_msg_verb(2)
    compute_path(sys.argv[1], sys.argv[2], sys.argv[3])

else:
    general.error("Invalid command line")
    
    

File: lab6/java/comp26120/simple_test_graph.java
package comp26120;

public class simple_test_graph {
    general settings;

    public simple_test_graph(general settings) {
	this.settings = settings;
    }

    public void compute_path(String algo, String src, String dst) {
	node_t s = new node_t(Integer.parseInt(src));
	node_t d = new node_t(Integer.parseInt(dst));

	graph_t g = new graph_t(10);
	g.graph_add_edge(new node_t(0), new weight_t(1), new node_t(1));
	g.graph_add_edge(new node_t(1), new weight_t(2), new node_t(2));
	g.graph_add_edge(new node_t(2), new weight_t(3), new node_t(3));
	g.graph_add_edge(new node_t(3), new weight_t(1), new node_t(4));
        g.graph_add_edge(new node_t(5), new weight_t(1), new node_t(6));
	g.graph_add_edge(new node_t(6), new weight_t(1), new node_t(8));
	g.graph_add_edge(new node_t(8), new weight_t(2), new node_t(9));
	g.graph_add_edge(new node_t(5), new weight_t(9), new node_t(9));
	if (algo.equals("bellman-ford-neg")) {
	    g.graph_add_edge(new node_t(3), new weight_t(-1), new node_t(0));
	    g.graph_add_edge(new node_t(6), new weight_t(-5), new node_t(7));
	} else {
	    g.graph_add_edge(new node_t(3), new weight_t(4), new node_t(0));
	    g.graph_add_edge(new node_t(5), new weight_t(5), new node_t(7));
	}

	weight_t[] h = new weight_t[g.graph_get_num_nodes()];
	for (int u = 0; u< g.graph_get_num_nodes(); ++ u) {
	    h[u] = new weight_t(java.lang.Math.abs(d.i - u));
	}

	if (algo.equals("bellman-ford")) {
	    sp_algorithms.bellman_ford(g, s).sssp_to_sp_result(d).print_sp_result(System.out);
	} else if (algo.equals("bellman-ford-neg")) {
	    sp_algorithms.bellman_ford(g, s).sssp_to_sp_result(d).print_sp_result(System.out);
	} else if (algo.equals("dijkstra")) {
	    sp_algorithms.dijkstra(g, s, d).sssp_to_sp_result(d).print_sp_result(System.out);
	} else if (algo.equals("astar")) {
	    sp_algorithms.astar_search(g,s,d,h).print_sp_result(System.out);
	} else if (algo.equals("bfs")) {
	    sp_algorithms.bfs(g,s,d).sssp_to_sp_result(d).print_sp_result(System.out);
	} else {
	    sp_algorithms.error ("Invalid algorithm name: " + algo);
	}

    }

    public static void main(String[] args) {
	general general_settings = new general();
	general_settings.set_msg_verb(-1);

	simple_test_graph ap = new simple_test_graph(general_settings);

	if (args.length == 3 ) {
	    ap.compute_path(args[0],args[1],args[2]);
	} else if (args.length == 4) {
	    if (args[3].equals("-v")) {
                general_settings.set_msg_verb(0);
            } else if (args[3].equals("-vv")) {
                general_settings.set_msg_verb(1);
            } else if (args[3].equals("-vvv")) {
                general_settings.set_msg_verb(2);
	    }

	    ap.compute_path(args[0],args[1],args[2]);
	} else {
	    sp_algorithms.error("Invalid command line");
	}
    }
}
